### PR TITLE
feat: bootstrapping tests

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
## Summary of changes

Changes introduced in this pull request:

- follow-up for https://github.com/ChainSafe/forest/pull/4580 and part of https://github.com/ChainSafe/forest/issues/4576
- adds bootstrapping tests in the CI so that the bootstrapping capability is not broken in the future.
- I had to split it into multiple jobs due to GH worker restrictions (we were running out of space, only 14G of disk space and the snapshots itself is 4G, which cannot be used directly by Lotus...).

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by LesnyRumcajs** [comment]
Likely out of disk space, may need to tinker around given a 4 GB snapshot is tossed around on a 14 GB machine. Or move to a paid runner.

File: scripts/tests/bootstrapper/lotus.dockerfile [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+ARG LOTUS_VERSION
+RUN git clone --depth 1 --branch ${LOTUS_VERSION} https://github.com/filecoin-project/lotus.git .
+
+# Replace the default bootstrap peers with a pre-defined one. This is needed on this level because the bootstrap peers are compiled into the binary and not configurable at runtime.
+ARG BOOTSTRAPPER
+RUN echo ${BOOTSTRAPPER} > build/bootstrap/calibnet.pi
+
+# Apply a patch to Lotus so that local IPs are not discarded
+COPY patch-lotus.diff .
```
Comment (hanabi1224) [inline scripts/tests/bootstrapper/lotus.dockerfile:21]: We could get rid of this custom dockerfile once https://github.com/filecoin-project/lotus/pull/12289 is merged and released.

File: scripts/tests/bootstrapper/lotus.dockerfile [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+COPY --from=lotus-builder /usr/lib/*/libltdl.so.7   /lib/
+COPY --from=lotus-builder /usr/lib/*/libnuma.so.1   /lib/
+COPY --from=lotus-builder /usr/lib/*/libhwloc.so.*  /lib/
+COPY --from=lotus-builder /usr/lib/*/libOpenCL.so.1 /lib/
+
+# Copy only the binaries relevant for the bootstrap test
+COPY --from=lotus-builder /lotus/lotus /usr/local/bin/
+
+WORKDIR /lotus
+
```
Comment (hanabi1224) [inline scripts/tests/bootstrapper/lotus.dockerfile:49]: Should we set `ENV LOTUS_SYNC_BOOTSTRAP_PEERS 1` in the dockerfile to align with the patched bootstrap file.

File: scripts/tests/bootstrapper/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    volumes:
+      - node-data:/data
+      - lotus-data:/var/lib/lotus
+      - filecoin-proofs:${FIL_PROOFS_PARAMETER_CACHE}
+    networks:
+      - bootstrap-tests
+    environment:
+      - FIL_PROOFS_PARAMETER_CACHE=${FIL_PROOFS_PARAMETER_CACHE}
+      - LOTUS_API_LISTENADDRESS=/ip4/0.0.0.0/tcp/${LOTUS_RPC_PORT}/http
+      - LOTUS_SYNC_BOOTSTRAP_PEERS=1
```
Comment (hanabi1224) [inline scripts/tests/bootstrapper/docker-compose.yml:142]: This is not needed if it lives in the Dockerfile

File: scripts/tests/bootstrapper/lotus.dockerfile (hunk 2 above)
Comment (LesnyRumcajs) [reply scripts/tests/bootstrapper/lotus.dockerfile:49]: Yes, this makes sense.

File: scripts/tests/bootstrapper/lotus.dockerfile [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --no-modify-path --profile minimal
+
+ENV PATH="/root/.cargo/bin:${PATH}"
+
+ARG LOTUS_VERSION
+RUN git clone --depth 1 --branch ${LOTUS_VERSION} https://github.com/filecoin-project/lotus.git .
+
+# Replace the default bootstrap peers with a pre-defined one. This is needed on this level because the bootstrap peers are compiled into the binary and not configurable at runtime.
+ARG BOOTSTRAPPER
+RUN echo ${BOOTSTRAPPER} > build/bootstrap/calibnet.pi
```
Comment (hanabi1224) [inline scripts/tests/bootstrapper/lotus.dockerfile:18]: We can get rid of this custom Dockerfile if https://github.com/filecoin-project/lotus/pull/12292 is approved

Comment (LesnyRumcajs) [reply scripts/tests/bootstrapper/lotus.dockerfile:18]: And released, so let's do it then. :)

**Review by hanabi1224 — APPROVED** [review]

File: scripts/tests/bootstrapper/lotus.dockerfile [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+RUN CGO_CFLAGS_ALLOW="-D__BLST_PORTABLE__" \
+    CGO_CFLAGS="-D__BLST_PORTABLE__" \
+    FFI_USE_BLST_PORTABLE="1" \
+    FFI_USE_GPU="0" \
+    make calibnet && strip lotus*
+
+FROM ubuntu:22.04
+
+# Need to copy the relevant shared libraries from the builder image.
+# See https://github.com/filecoin-project/lotus/blob/master/Dockerfile
```
Comment (aatifsyed) [inline scripts/tests/bootstrapper/lotus.dockerfile:34]: I'd prefer a permalink to a tagged release

Comment (LesnyRumcajs) [reply scripts/tests/bootstrapper/lotus.dockerfile:34]: It's not useful in the context of a constantly updated tag. Do you agree that bumping the Lotus version only in the environment file is more maintainable that going through all the linked permalinks?

**Review by ruseinov — APPROVED** [review]

**Pull request merged**